File: geode-client-native/src/tests/clicache/TestObject/DeltaPSTObject.cs
//=========================================================================
// Copyright (c) 2002-2014 Pivotal Software, Inc. All Rights Reserved.
// This product is protected by U.S. and international copyright
// and intellectual property laws. Pivotal products are covered by
// more patents listed at http://www.pivotal.io/patents.
//========================================================================

using System;

namespace GemStone.GemFire.Cache.Tests
{
  public class DeltaPSTObject
    : IGFSerializable, IGFDelta
  {
    private long timestamp;
    private Int32 field1;
    private byte field2;
    private CacheableBytes valueData;
    private bool hasDelta = false;

    public DeltaPSTObject()
    {
      timestamp = 0;
      valueData = null;
    }

    public DeltaPSTObject(Int32 size, bool encodeKey, bool encodeTimestamp)
    {
      DateTime startTime = DateTime.Now;
      timestamp = startTime.Ticks * (1000000 / TimeSpan.TicksPerMillisecond);
      field1 = 1234;
      field2 = 123;
      if (size == 0)
      {
        valueData = null;
      }
      else
      {
        encodeKey = true;
        valueData = ArrayOfByte.Init(size, encodeKey, false);
      }
    }
    public UInt32 ObjectSize
    {
      get
      {
        return 0;
      }
    }
    public UInt32 ClassId
    {
      get
      {
        return 42;
      }
    }
    public IGFSerializable FromData(DataInput input)
    {
      timestamp = input.ReadInt64();
      field1 = input.ReadInt32();
      field2 =input.ReadByte();
      valueData = (CacheableBytes)input.ReadObject();
      return this;
    }
    public void ToData(DataOutput output)
    {
      output.WriteInt64(timestamp);
      output.WriteInt32(field1);
      output.WriteByte(field2);
      output.WriteObject(valueData);
    }
    public void ToDelta(DataOutput output)
    {
      output.WriteInt32(field1);
      output.WriteInt64(timestamp);
    }
    public void FromDelta(DataInput input)
    {
      field1 = input.ReadInt32();
      timestamp = input.ReadInt64();
    }
    public bool HasDelta()
    {
      return hasDelta;
    }

    public void Update()
    {
      IncrementField1();
      ResetTimestamp();
    }
    public void IncrementField1()
    {
      ++field1;
    }
    public void ResetTimestamp()
    {
      DateTime startTime = DateTime.Now;
      timestamp = startTime.Ticks * (1000000 / TimeSpan.TicksPerMillisecond);
    }
    public Object Clone()
    {
      return new DeltaPSTObject();
    }
    public static IGFSerializable CreateDeserializable()
    {
      return new DeltaPSTObject();
    }
  }
}

File: src/tests/cpp/testobject/DeltaPSTObject.cpp
/*=========================================================================
 * Copyright (c) 2010-2014 Pivotal Software, Inc. All Rights Reserved.
 * This product is protected by U.S. and international copyright
 * and intellectual property laws. Pivotal products are covered by
 * one or more patents listed at http://www.pivotal.io/patents.
 *=========================================================================
 */
#include "fwklib/GsRandom.hpp"
#include "ArrayOfByte.hpp"
#include "DeltaPSTObject.hpp"

using namespace gemfire;
using namespace testframework;
using namespace testobject;

DeltaPSTObject::DeltaPSTObject(int size, bool encodeKey, bool encodeTimestamp) {
  ACE_Time_Value startTime;
  startTime = ACE_OS::gettimeofday();
  ACE_UINT64 tusec = 0;
  startTime.to_usec(tusec);
  timestamp = tusec * 1000;
  field1 = 1234;
  field2 = '*';
  if (size == 0) {
    valueData = NULLPTR;
  } else {
    encodeKey = true;
    valueData = ArrayOfByte::init(size, encodeKey, false);
  }
}
void DeltaPSTObject::fromDelta(DataInput& input) {
  input.readInt(&field1);
  input.readInt(reinterpret_cast<int64_t*>(&timestamp));
}
void DeltaPSTObject::toDelta(DataOutput& output) const {
  output.writeInt(static_cast<int32_t>(field1));
  output.writeInt(static_cast<int64_t>(timestamp));
}
void DeltaPSTObject::toData(gemfire::DataOutput& output) const {
  output.writeInt(static_cast<int64_t>(timestamp));
  output.writeInt(static_cast<int32_t>(field1));
  output.write(field2);
  output.writeObject(valueData);
}

gemfire::Serializable* DeltaPSTObject::fromData(gemfire::DataInput& input) {
  input.readInt(reinterpret_cast<int64_t*>(&timestamp));
  input.readInt(&field1);
  input.read(&field2);
  input.readObject(valueData);
  return this;
}
CacheableStringPtr DeltaPSTObject::toString() const {
  char buf[102500];
  sprintf(
      buf,
      "DeltaPSTObject:[timestamp = %lld field1 = %d field2 = %c valueData=%d ]",
      timestamp, field1, field2, valueData->length());
  return CacheableString::create(buf);
}
